### PR TITLE
Add support for pickling exceptions

### PR DESCRIPTION
**Pull request by kyleknap**
There has been a few requests/PRs for this:

https://github.com/boto/botocore/issues/834
https://github.com/boto/botocore/pull/1191
https://github.com/boto/botocore/pull/835

### CODE REVIEW

File: botocore/exceptions.py [hunk 1]
```diff
@@ -395,6 +414,12 @@ def _get_retry_info(self, response):
                                   metadata['RetryAttempts'])
         return retry_info
 
+    def __reduce__(self):
+        # Subclasses of ClientError's are dynamically generated and
+        # cannot be pickled unless they are attributes of a
+        # module. So at the very least return a ClientError back.
+        return ClientError, (self.response, self.operation_name)
```
Comment (joguSD) [inline botocore/exceptions.py]: This is a sub class of `BotoCoreError` and not `ClientError`.

Comment (JordonPhillips) [reply botocore/exceptions.py]: This ***is*** `ClientError`

**Comment by kyleknap** [comment]
@JordonPhillips Yeah it is unfortunate. I would not say there is nothing we can do about generated exceptions. We can potentially still add it in the future, but it may be tricky. For example, this PR: https://github.com/boto/botocore/pull/1191 has an example of how you could do it. The only hesitation that I had in looking it over was that PR was it set dynamic exception modules as attributes to the global `exceptions` module when these dynamically generated exceptions are only scoped to client instances.

However just making the exceptions pickleable goes a long way in using `botocore` in processes. The fact that most of the errors we raise in botocore are not pickleable is pretty annoying.

File: botocore/exceptions.py (hunk 1 above)
Comment (joguSD) [reply botocore/exceptions.py]: Oh, for some reason I though this was in `ConnectionError` and was really confused.

**Pull request merged**